jupyter_collaboration/handlers.py

```diff
@@ -149,130 +186,33 @@ async def __anext__(self):
-            model = await ensure_async(
-                self.contents_manager.get(file_path, type=file_type, format=file_format)
-            )
-            self.last_modified = model["last_modified"]
-            # check again if ready, because loading the file can be async
-            if not self.room.ready:
-                # try to apply Y updates from the YStore for this document
-                read_from_source = True
-                if self.room.ystore is not None:
-                    try:
-                        await self.room.ystore.apply_updates(self.room.ydoc)
-                        read_from_source = False
-                    except YDocNotFound:
-                        # YDoc not found in the YStore, create the document from the source file (no change history)
-                        pass
-                if not read_from_source:
-                    # if YStore updates and source file are out-of-sync, resync updates with source
-                    if self.room.document.source != model["content"]:
-                        read_from_source = True
-
-                if read_from_source:
-                    self.room.document.source = model["content"]
-                    if self.room.ystore:
-                        await self.room.ystore.encode_state_as_update(self.room.ydoc)
-                self.room.document.dirty = False
-                self.room.ready = True
-                self.room.watcher = asyncio.create_task(self.watch_file())
-                # save the document when changed
-                self.room.document.observe(self.on_document_change)
-
-    async def watch_file(self):
-        assert isinstance(self.room, DocumentRoom)
-        poll_interval = self.settings["collaborative_file_poll_interval"]
-        if not poll_interval:
-            self.room.watcher = None
-            return
-        while True:
-            await asyncio.sleep(poll_interval)
-            await self.maybe_load_document()
-
-    async def maybe_load_document(self):
-        assert isinstance(self.room, DocumentRoom)
-        file_format, file_type, file_path = self.get_file_info()
-        async with self.lock:
-            model = await ensure_async(
-                self.contents_manager.get(
-                    file_path, content=False, type=file_type, format=file_format
-                )
-            )
-        # do nothing if the file was saved by us
-        if self.last_modified < model["last_modified"]:
-            self.log.debug("Reverting file that had out-of-band changes: %s", file_path)
-            model = await ensure_async(
-                self.contents_manager.get(file_path, type=file_type, format=file_format)
-            )
-            self.room.document.source = model["content"]
-            self.last_modified = model["last_modified"]
+            # Initialize the room
+            async with self.room.lock:
+                if not self.room.ready:
+                    await self.room.initialize()
```

Comment:
I think the room lock should not be used outside of the room, but rather in its `initialize()` method.
Maybe also the check for room ready.